NPE in v1.7.5

Not quite sure what's going on here.   Either MainDrawerActivity can't resume or we're getting a NPE at:

``` java
at org.mozilla.mozstumbler.client.util.NotificationUtil.update(NotificationUtil.java:79)
```

Stacktrace:

``` java
java.lang.RuntimeException: Unable to resume activity {org.mozilla.mozstumbler/org.mozilla.mozstumbler.client.navdrawer.MainDrawerActivity}: java.lang.NullPointerException
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2919)
at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2948)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2354)
at android.app.ActivityThread.access$700(ActivityThread.java:159)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1316)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:176)
at android.app.ActivityThread.main(ActivityThread.java:5419)
at java.lang.reflect.Method.invokeNative(Native Method)
at java.lang.reflect.Method.invoke(Method.java:525)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
at dalvik.system.NativeStart.main(Native Method)
Caused by: java.lang.NullPointerException
at android.os.Parcel.readException(Parcel.java:1437)
at android.os.Parcel.readException(Parcel.java:1385)
at android.app.INotificationManager$Stub$Proxy.enqueueNotificationWithTag(INotificationManager.java:338)
at android.app.NotificationManager.notify(NotificationManager.java:139)
at android.app.NotificationManager.notify(NotificationManager.java:112)
at org.mozilla.mozstumbler.client.util.NotificationUtil.update(NotificationUtil.java:79)
at org.mozilla.mozstumbler.client.util.NotificationUtil.updateLastUploadedLabel(NotificationUtil.java:112)
at org.mozilla.mozstumbler.client.navdrawer.MetricsView.updateLastUploadedLabel(MetricsView.java:263)
at org.mozilla.mozstumbler.client.navdrawer.MetricsView.updateSentStats(MetricsView.java:282)
at org.mozilla.mozstumbler.client.navdrawer.MetricsView.update(MetricsView.java:222)
at org.mozilla.mozstumbler.client.navdrawer.MainDrawerActivity.onPostResume(MainDrawerActivity.java:203)
at android.app.Activity.performResume(Activity.java:5463)
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2909)
... 12 more
java.lang.NullPointerException
at android.os.Parcel.readException(Parcel.java:1437)
at android.os.Parcel.readException(Parcel.java:1385)
at android.app.INotificationManager$Stub$Proxy.enqueueNotificationWithTag(INotificationManager.java:338)
at android.app.NotificationManager.notify(NotificationManager.java:139)
at android.app.NotificationManager.notify(NotificationManager.java:112)
at org.mozilla.mozstumbler.client.util.NotificationUtil.update(NotificationUtil.java:79)
at org.mozilla.mozstumbler.client.util.NotificationUtil.updateLastUploadedLabel(NotificationUtil.java:112)
at org.mozilla.mozstumbler.client.navdrawer.MetricsView.updateLastUploadedLabel(MetricsView.java:263)
at org.mozilla.mozstumbler.client.navdrawer.MetricsView.updateSentStats(MetricsView.java:282)
at org.mozilla.mozstumbler.client.navdrawer.MetricsView.update(MetricsView.java:222)
at org.mozilla.mozstumbler.client.navdrawer.MainDrawerActivity.onPostResume(MainDrawerActivity.java:203)
at android.app.Activity.performResume(Activity.java:5463)
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2909)
at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2948)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2354)
at android.app.ActivityThread.access$700(ActivityThread.java:159)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1316)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:176)
at android.app.ActivityThread.main(ActivityThread.java:5419)
at java.lang.reflect.Method.invokeNative(Native Method)
at java.lang.reflect.Method.invoke(Method.java:525)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
at dalvik.system.NativeStart.main(Native Method)
```

--- a/android/src/main/java/org/mozilla/mozstumbler/client/util/NotificationUtil.java
+++ b/android/src/main/java/org/mozilla/mozstumbler/client/util/NotificationUtil.java
@@ -14,8 +14,16 @@ import org.mozilla.mozstumbler.client.MainApp;
 import org.mozilla.mozstumbler.client.navdrawer.MainDrawerActivity;
 import org.mozilla.mozstumbler.svclocator.ServiceLocator;
 import org.mozilla.mozstumbler.svclocator.services.ISystemClock;
+import org.mozilla.mozstumbler.svclocator.services.log.ILogger;
+import org.mozilla.mozstumbler.svclocator.services.log.LoggerUtil;
 
 public class NotificationUtil {
+
+    private static final ILogger Log = (ILogger) ServiceLocator
+                                                    .getInstance()
+                                                    .getService(ILogger.class);
+    private static final String LOG_TAG = LoggerUtil.makeLogTag(NotificationUtil.class);
+
     public static final int NOTIFICATION_ID = 1;
     private static final long UPDATE_FREQUENCY = 60 * 1000;
     private static String sStopTitle;
@@ -29,7 +37,10 @@ public class NotificationUtil {
     }
 
     private Notification build() {
-        PendingIntent turnOffIntent = PendingIntent.getBroadcast(mContext, 0, new Intent(MainApp.INTENT_TURN_OFF), 0);
+        PendingIntent turnOffIntent = PendingIntent.getBroadcast(mContext,
+                                                                0,
+                                                                new Intent(MainApp.INTENT_TURN_OFF),
+                                                                0);
 
         Intent notificationIntent = new Intent(mContext, MainDrawerActivity.class);
         notificationIntent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_FROM_BACKGROUND);
@@ -73,10 +84,17 @@ public class NotificationUtil {
                 .build();
     }
 
-    private void update() {
+    void update() {
         Notification notification = build();
         NotificationManager nm = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
+        if (nm == null) {
+            // This really *shouldn't* happen, but we get this on some devices. See :
+            // https://github.com/mozilla/MozStumbler/issues/1564
+            Log.d(LOG_TAG, "Couldn't acquire notification service");
+            return;
+        }
         nm.notify(NOTIFICATION_ID, notification);
+
     }
 
     public Notification buildNotification(String stopTitle) {

--- a/android/src/test/java/org/mozilla/mozstumbler/client/QueuedCountsTrackerTest.java
+++ b/android/src/test/java/org/mozilla/mozstumbler/client/QueuedCountsTrackerTest.java
@@ -18,11 +18,8 @@ import org.robolectric.Robolectric;
 import org.robolectric.RobolectricTestRunner;
 import org.robolectric.annotation.Config;
 
-import static android.test.MoreAsserts.assertNotEqual;
 import static junit.framework.Assert.assertEquals;
 import static junit.framework.Assert.assertNotNull;
-import static junit.framework.Assert.assertNull;
-import static junit.framework.Assert.assertTrue;
 import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.spy;
 

--- a/android/src/test/java/org/mozilla/mozstumbler/client/util/NotificationUtilTest.java
+++ b/android/src/test/java/org/mozilla/mozstumbler/client/util/NotificationUtilTest.java
@@ -1,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.mozstumbler.client.util;
+
+
+import android.content.Context;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@Config(emulateSdk = 18)
+@RunWith(RobolectricTestRunner.class)
+public class NotificationUtilTest {
+
+    @Test
+    public void testNotificationUpdateNPESafe() {
+        // NoticationUtil should not die if the NotificationManager comes back as null
+
+        Context ctx = Robolectric.application;
+        ctx = spy(ctx);
+        doReturn(null).when(ctx).getSystemService(Context.NOTIFICATION_SERVICE);
+
+        NotificationUtil nu = new NotificationUtil(ctx);
+        nu = spy(nu);
+
+        final long ARBITRARY_TIME = 555;
+        nu.updateLastUploadedLabel(ARBITRARY_TIME);
+        verify(ctx, times(1)).getSystemService(Context.NOTIFICATION_SERVICE);
+        verify(nu, times(1)).update();
+    }
+
+
+}

